zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -0,0 +1,240 @@
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
+      // write bound, but a drop in concurrency won't necessarily help.
+      limitListener.onIgnore();
+      throw e;
+    }
+  }
+
+  @Override
+  public void enqueue(Callback<V> callback) {
+    try {
+      executor.execute(new QueuedCall(callback));
+    } catch (Exception e) {
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
+      // write bound, but a drop in concurrency won't necessarily help.
+      limitListener.onIgnore();
+      throw e;
+    }
+  }
+
+  @Override
+  public void cancel() {
+    canceled = true;
+    if (call != null) {
+      call.cancel();
+    }
+  }
+
+  @Override
+  public boolean isCanceled() {
+    return canceled || (call != null && call.isCanceled());
+  }
+
+  @Override
+  public Call<V> clone() {
+    return new ThrottledCall<>(this);
+  }
+
+  /**
+   * @param name New name for the current Thread
+   * @return Previous name of the current Thread
+   */
+  static String setCurrentThreadName(String name) {
+    Thread thread = Thread.currentThread();
+    String originalName = thread.getName();
+    try {
+      thread.setName(name);
+      return originalName;
+    } catch (SecurityException e) {
+      return originalName;
+    }
+  }
+
+  final class QueuedCall implements Runnable {
+    final Callback<V> callback;
+
+    public QueuedCall(Callback<V> callback) {
+      this.callback = callback;
+    }
+
+    @Override
+    public void run() {
+      try {
+        if (canceled) {
+          return;
+        }
+
+        call = ThrottledCall.this.delegate.get();
+
+        String oldName = setCurrentThreadName(call.toString());
+        try {
+          enqueueAndWait();
+        } finally {
+          setCurrentThreadName(oldName);
+        }
+      } catch (Exception e) {
+        limitListener.onIgnore();
+        callback.onError(e);
+      }
+    }
+
+    void enqueueAndWait() {
+      ThrottledCallback<V> throttleCallback = new ThrottledCallback<>(callback, limitListener);
+      call.enqueue(throttleCallback);
+
+      // Need to wait here since the callback call will run asynchronously also.
+      // This ensures we don't exceed our throttle/queue limits.
+      throttleCallback.await();
+    }
+  }
+
+  static final class ThrottledCallback<V> implements Callback<V> {
+    final Callback<V> delegate;
+    final Listener limitListener;
+    final CountDownLatch latch;
+
+    public ThrottledCallback(Callback<V> delegate, Listener limitListener) {
+      this.delegate = delegate;
+      this.limitListener = limitListener;
+      this.latch = new CountDownLatch(1);
+    }
+
+    public void await() {
+      try {
+        latch.await();
+      } catch (InterruptedException e) {
+        limitListener.onIgnore();
+        throw new RuntimeException("Interrupted while blocking on a throttled call", e);
```

Comment:
thread.currentThread.setInterrupted or something I think is the pattern on catching IE